feat(docs-infra): implement popup to inform about the use of cookies

gkalpak opened the pull request:
This commit adds a popup to angular.io to inform the user about the use of cookies. Once the user confirms having read the info, the popup will not be shown on subsequent visits.

This commit is partly based on angular/material.angular.io#988.

Fixes #42209.

mary-poppins commented:
You can preview 2f21c6b at https://pr42259-2f21c6b.ngbuilds.io/.

jelbourn commented:
@gkalpak while chatting with @crisbeto about this for material.angular.io, some things to note:
* The pop-up shouldn't use a snackbar, because:
    1. The cookie prompt need to be the first thing in the tab order so that people using assistive technology encounter it first (even before the skip link)
    2. The prompt should be able to be dismissed without explicitly hitting okay (another snackbar opening would otherwise dismiss it)
    3. The text doesn't need to be announced via aria-live

Instead, I think the best way to do this is with regular div at the beginning of the tab order with `position: absolute` or `position: fixed`.

jelbourn commented:
Ahh, that was a typo, the prompt should *not* be dismissable without hitting okay

mary-poppins commented:
You can preview d4ea783 at https://pr42259-d4ea783.ngbuilds.io/.

mary-poppins commented:
You can preview fddbbe6 at https://pr42259-fddbbe6.ngbuilds.io/.

mary-poppins commented:
You can preview 46c5965 at https://pr42259-46c5965.ngbuilds.io/.

mary-poppins commented:
You can preview 633945e at https://pr42259-633945e.ngbuilds.io/.

mary-poppins commented:
You can preview 539dcee at https://pr42259-539dcee.ngbuilds.io/.

petebacondarwin commented on aio/src/app/shared/storage.service.spec.ts:
```diff
@@ -0,0 +1,40 @@
+      localStorage: jasmine.createSpy('get localStorage'),
+      sessionStorage: jasmine.createSpy('get sessionStorage'),
+    };
+    instance = new StorageService(Object.defineProperties({} as Window, {
+      localStorage: {get: storageGetterSpies.localStorage},
+      sessionStorage: {get: storageGetterSpies.sessionStorage},
+    }));
```
I think I would find this a lot easier to grok if there was a `MockWindow` object that is referenced rather than this inline definition.

mary-poppins commented:
You can preview e6bba4e at https://pr42259-e6bba4e.ngbuilds.io/.

IgorMinar commented on aio/src/app/layout/cookies-popup/cookies-popup.component.ts:
```diff
@@ -0,0 +1,35 @@
+  hasAcceptedCookies: boolean;
+
+  constructor(@Inject(LocalStorage) private storage: Storage) {
+    this.hasAcceptedCookies = this.storage.getItem(storageKey) === 'true';
```
maybe we could lazy load this component only if the key is not found in the local storage? that way we don't need to fetch this code up front since for 9x% of all sessions this will be just dead code.

we could implement that by moving this check into a standalone service that checks the local storage, and if the key is not found it fetches the component and dynamically instantiates it.

petebacondarwin replied:
I believe we could do this already by using the `LayCustomElementComponent` in an `ngIf`:

```
<aio-lazy-ce selector="aio-cookies-popup" *ngIf="!hasAcceptedCookies"></aio-lazy-ce>
```

Then we just need to move the `hasAcceptedCookies` property to `AppComponent`.

IgorMinar reviewed (commented):
This looks quite good to me, thanks @gkalpak for putting this together. Just a few small issues:

- consider lazy loading the component
- when testing the preview I don't see any visual indication of either of the two popup buttons having a focus. Can you please take a look at that?

petebacondarwin reviewed (commented):
LGTM - I agree with @IgorMinar about trying to make this lazily loaded.

jelbourn commented on aio/src/app/layout/cookies-popup/cookies-popup.component.ts:
```diff
@@ -0,0 +1,35 @@
+
+      <div class="actions">
+        <a mat-button href="https://policies.google.com/technologies/cookies" target="_blank" rel="noopener">
+          LEARN MORE
```
The buttons' text should be normal case with `text-transform` to capitalize, otherwise screen-reader can read them out as "L - E - A - R - N..."

jessicajaniuk reviewed (commented):
This adds 3mb in size to the main bundle, and I'm very surprised by that given the scope of this change.

Edit: My brain apparently forgot that kilobytes exist.

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

(the delta is 3kb which looks reasonable given the code added to the app)

jessicajaniuk reviewed (approved):
LGTM 🍪 

reviewed-for: size-tracking

gkalpak committed:
build(docs-infra): align code with new CLI v12 apps

This commit aligns the angular.io config files more closely to how a
newly generated CLI v12 app would look like. This helps validate the
setup and makes it easier to apply new chages in the future (by
preventing the angular.io layout from deviating too much from the
default new app layout).

gkalpak committed:
test(docs-infra): fix warnings in `ThemeToggleComponent` unit tests.

due to unknown `<mat-icon>` element

This commit fixes some warnings in the unit tests of the
`ThemeToggleComponent`, which were caused by the following:

- The `<mat-icon>` element used in `ToggleThemeComponent`'s template was
  not declared in tests.
- The `dark-theme.css` and `light-theme.css` files requested by
  `ToggleThemeComponent` were not available.

gkalpak committed:
refactor(docs-infra): provide `local-/sessionStorage` via DI

Previously, we had the same logic in a couple of places to safely access
the `Window`'s `local-/sessionStorage` and provide a no-op fallback if
necessary. Soon, we will need the same logic for the cookies popup
(see #42209).

This commit reduces code duplication by providing
`local-/sessionStorage` as injectables and sharing the logic for
accessing them safely. This also makes it easier to mock the storage in
tests without having to mess with the actual `Window` object.

NOTE:
This commit actually decreases the payload size in the `main` bundle by
40B.

gkalpak committed:
feat(docs-infra): implement popup to inform about the use of cookies

This commit adds a popup to angular.io to inform the user about the use
of cookies. Once the user confirms having read the info, the popup will
not be shown on subsequent visits.

This commit is partly based on angular/material.angular.io#988.

Fixes #42209

gkalpak committed:
fixup! feat(docs-infra): implement popup to inform about the use of cookies

mary-poppins commented:
You can preview 246e038 at https://pr42259-246e038.ngbuilds.io/.

gkalpak commented:
As discussed "offline", let's merge this as is and investigate whether lazy-loading can offer a good experience wrt a11y as a follow-up.
